### ISSUE TITLE
Saving and accessing loss values in hk.Modules

### ISSUE THREAD

**Issue by njunge94** [deepmind/dm-haiku#124]
Hello dear Haiku community! I just started to use a stack of JAX, Optax and Haiku for a small personal project concerning neural style transfer, so I don't have a lot of experience right now. I am trying to implement the algorithm from this Pytorch tutorial: https://pytorch.org/tutorials/advanced/neural_style_tutorial.html

There, two new modules are introduced as identity layers that log two types of loss values between an input image and a reference image. As far as I've understood, this cannot be ported directly to JAX/Haiku, since these values are unavailable in a pure function returning the net coming from `hk.transform`. I believe the Pytorch code in that tutorial should be equivalent to this:

```
class StyleLoss(hk.Module):

    def __init__(self, target, name: Optional[str] = None):
        super(StyleLoss, self).__init__(name=name)
        self.target_g = jax.lax.stop_gradient(gram_matrix(target))
        self.loss = None

    def __call__(self, x):
        g = gram_matrix(x)
        self.loss = jnp.mean(jnp.square(g - self.target_g))
        return x
```
With some fixed parameters built in (hence the `stop_gradient`).

How would I need to change this code to be able to access these loss values that are updated in the call method in the `params` of a training loop for computing gradients, such as the one in the MNIST example? I have tried using `hk.get_parameter`, but unsuccessfully so far. Thank you in advance for some direction!

**Comment by tomhennigan** [comment]
Hi @njunge94,  in this case I would suggest looking at `hk.set_state` which allows you to pass intermediate values (in and) out of transformed functions. You may also be interested in looking at [oryx](https://www.tensorflow.org/probability/oryx/notebooks/a_tour_of_oryx) which has a general reap/sow mechanism that works anywhere in JAX.

I've put two examples together here: https://colab.research.google.com/gist/tomhennigan/796e5be7b41370d43032d2e74da10116/examples-of-saving-intermediate-values.ipynb

**Comment by njunge94** [comment]
Thank you so much for the quick response @tomhennigan! That is *exactly* what I was looking for, down to the last detail. I am closing this then as resolved, if there are no objections.
